src/network/address.rs

```diff
@@ -191,5 +365,113 @@ mod test {
         let addr = Address::new(&onionaddr, ServiceFlags::NONE);
         assert!(addr.socket_addr().is_err());
     }
-}
 
+    #[test]
+    fn serialize_addrv2ip_test() {
+        // Taken from https://github.com/bitcoin/bitcoin/blob/12a1c3ad1a43634d2a98717e49e3f02c4acea2fe/src/test/net_tests.cpp#L348
+
+        let ip = AddrV2IP::IPV4(Ipv4Addr::new(1, 2, 3, 4));
+        assert_eq!(serialize(&ip), Vec::from_hex("010401020304").unwrap());
+
+        let ip = AddrV2IP::IPV6(Ipv6Addr::from_str("1a1b:2a2b:3a3b:4a4b:5a5b:6a6b:7a7b:8a8b").unwrap());
+        assert_eq!(serialize(&ip), Vec::from_hex("02101a1b2a2b3a3b4a4b5a5b6a6b7a7b8a8b").unwrap());
+
+        let ip = AddrV2IP::TorV2(FromHex::from_hex("f1f2f3f4f5f6f7f8f9fa").unwrap());
+        assert_eq!(serialize(&ip), Vec::from_hex("030af1f2f3f4f5f6f7f8f9fa").unwrap());
+
+        let ip = AddrV2IP::TorV3(FromHex::from_hex("53cd5648488c4707914182655b7664034e09e66f7e8cbf1084e654eb56c5bd88").unwrap());
+        assert_eq!(serialize(&ip), Vec::from_hex("042053cd5648488c4707914182655b7664034e09e66f7e8cbf1084e654eb56c5bd88").unwrap());
+
+        let ip = AddrV2IP::I2P(FromHex::from_hex("a2894dabaec08c0051a481a6dac88b64f98232ae42d4b6fd2fa81952dfe36a87").unwrap());
+        assert_eq!(serialize(&ip), Vec::from_hex("0520a2894dabaec08c0051a481a6dac88b64f98232ae42d4b6fd2fa81952dfe36a87").unwrap());
+
+        let ip = AddrV2IP::CJDNS(Ipv6Addr::from_str("fc00:1:2:3:4:5:6:7").unwrap());
+        assert_eq!(serialize(&ip), Vec::from_hex("0610fc000001000200030004000500060007").unwrap());
+
+        let ip = AddrV2IP::Unknown(170, Vec::from_hex("01020304").unwrap());
+        assert_eq!(serialize(&ip), Vec::from_hex("aa0401020304").unwrap());
+    }
+
+    #[test]
+    fn deserialize_addrv2ip_test() {
+        // Taken from https://github.com/bitcoin/bitcoin/blob/12a1c3ad1a43634d2a98717e49e3f02c4acea2fe/src/test/net_tests.cpp#L386
+
+        // Valid IPv4.
+        let ip: AddrV2IP = deserialize(&Vec::from_hex("010401020304").unwrap()).unwrap();
+        assert_eq!(ip, AddrV2IP::IPV4(Ipv4Addr::new(1, 2, 3, 4)));
+
+        // Invalid IPv4, valid length but address itself is shorter.
+        assert!(deserialize::<AddrV2IP>(&Vec::from_hex("01040102").unwrap()).is_err());
+
+        // Invalid IPv4, with bogus length.
+        assert!(deserialize::<AddrV2IP>(&Vec::from_hex("010501020304").unwrap()).is_err());
+
+        // Invalid IPv4, with extreme length.
+        assert!(deserialize::<AddrV2IP>(&Vec::from_hex("01fd010201020304").unwrap()).is_err());
+
+        // Valid IPv6.
+        let ip: AddrV2IP = deserialize(&Vec::from_hex("02100102030405060708090a0b0c0d0e0f10").unwrap()).unwrap();
+        assert_eq!(ip, AddrV2IP::IPV6(Ipv6Addr::from_str("102:304:506:708:90a:b0c:d0e:f10").unwrap()));
```

Comment:
Missing "Valid IPv6. contains embedded "internal" ?

Reply:
It looked like internal addresses were an implementation detail of bitcoin core